controllers/capabilities/openapi_controller.go

```diff
@@ -358,6 +364,55 @@ func (r *OpenAPIReconciler) validateOpenAPIAs3scaleProduct(openapiCR *capabiliti
 	return nil
 }
 
+func (r *OpenAPIReconciler) validateOASExtensions(openapiObj *openapi3.T) error {
+	extensionErrors := field.ErrorList{}
+	productExtensionPath := field.NewPath("x-3scale-product")
+	metricsExtensionPath := productExtensionPath.Child("metrics")
+	policiesExtensionPath := productExtensionPath.Child("policies")
+
+	// Validate OAS root product extension
+	rootProductExtension, err := helper.NewOasRootProductExtension(openapiObj)
+	if err != nil {
+		return err
+	}
+
+	// Validate metrics
+	if rootProductExtension != nil && rootProductExtension.Metrics != nil {
+		// Loop through policies in extension and create PolicyConfig objects
+		for metricKey, metric := range rootProductExtension.Metrics {
+			if metric.Name == "" {
+				extensionErrors = append(extensionErrors, field.Required(metricsExtensionPath, fmt.Sprintf("metric %s is missing a friendlyName", metricKey)))
+			}
+			if metric.Unit == "" {
+				extensionErrors = append(extensionErrors, field.Required(metricsExtensionPath, fmt.Sprintf("metric %s is missing a unit", metricKey)))
+			}
+		}
+
+	}
+
+	// Validate policies
+	if rootProductExtension != nil && rootProductExtension.Policies != nil {
+		// Loop through policies in extension and create PolicyConfig objects
+		for _, policy := range rootProductExtension.Policies {
+			if policy.Name == "" {
+				extensionErrors = append(extensionErrors, field.Required(policiesExtensionPath, "one or more policies are missing a name"))
+			}
+			if policy.Version == "" {
+				extensionErrors = append(extensionErrors, field.Required(policiesExtensionPath, fmt.Sprintf("policy %s is missing a version", policy.Name)))
+			}
+		}
+	}
+
+	if len(extensionErrors) == 0 {
```

Comment:
Do we not need to validate the application plans?
When I created the oas secret without application plan name, the product CR responds with following error:
```
Invalid value: v1beta1.BackendUsageSpec{Path:"/"}: backend usage does not have valid backend reference., spec.applicationPlans[plan01].limits[0].metricMethodRef.backend: Invalid value: "Extended_Petstore": plan limit has invalid backend reference.]'
```
am I missing something?

Reply:
I initially skipped validating the application plans in the openapi_controller because this is already done by the product_controller [here](https://github.com/3scale/3scale-operator/blob/34e4398bb648e45af12a13795bf8cbf9676214c1/apis/capabilities/v1beta1/product_types.go#L1254) which then passes the error to the OpenAPI CR.

I can also validate application plans in the openapi_controller though which will prevent the broken Product CR from getting created in the first place.

Reply:
We actually don't need to perform the application plan validation in the openapi_controller in order to prevent broken Product CRs from getting created. 

This is because the product gets validated [here](https://github.com/3scale/3scale-operator/blob/573623e1976f87db295ebc84b19688684945f3de/controllers/capabilities/openapi_product_reconciler.go#L177) before it is created [here](https://github.com/3scale/3scale-operator/blob/573623e1976f87db295ebc84b19688684945f3de/controllers/capabilities/openapi_product_reconciler.go#L71). So what I think we should do is just enhance the [existing product validation](https://github.com/3scale/3scale-operator/blob/34e4398bb648e45af12a13795bf8cbf9676214c1/apis/capabilities/v1beta1/product_types.go#L1254-L1389) to check for the `invalid backend reference`.

Then the openapi_controller extension validation only checks for extension related issues and the product_controller validation handles everything else.